### PR TITLE
&duydl [MNT] Ensure Update Contributors does not run on main

### LINKED ISSUE #6181: [ENH] ensure `update-contributors` workflow does not run on `main`
(good first issue for contrinbutors with GHA experience)

The `update-contributors` workflow (which compiles the `all-contributorsrc` list into the `CONTRIBUTORS` page) is triggered on any branch, including `main`.

On `main`, however, it fails due to credentials, and anyway nothing should directly modify `main`.

An attempt to change the trigger condition failed: https://github.com/sktime/sktime/pull/6133, this seems due to trying to push to forks, as @yarnabrina has noted and @Xinyu-Wu-0000 explains here: https://github.com/sktime/sktime/pull/6093#issuecomment-2011074209

It would be great if we could modify the workflow so:

* it does not run on `main`
* it does not invalidate results of the normal CI which are no longer displayed in the GitHub UI. Optimally, the "update contributors" worflow (or a job) should run *before*, not *after* running normal CI.

### PR DESCRIPTION
**Pull request by Greyisheep**
#### Reference Issues/PRs
Fixes #6181.

#### What does this implement/fix? Explain your changes.
This implement ensures that `Update Contributors` does not run on `main`. It also makes sure `build wheels` only triggers when `release` is triggered and and `Update Contributors` is completed.

#### Did you add any tests for the change?
Not yet, but I am considering to add based on feedback and possible merge

#### Any other comments?
I think the workflow is too tightly woven together, and can be broken down further to ensure seamlessness.

### CODE REVIEW

**Comment by Greyisheep** [comment]
@fkiraly 

Please, check linked PR, and give feedback

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks! Very useful!

Few questions/remarks:

* I think you are accidentally adding empty spaces, linting will fail
* can you explain what the addition in the release workflow `wheels.yml` will do?
* I'd rather not couple the "update contributors" workflow to the release workflow, imo when the release button is being pushed, at that point we should assume that the contributors list is up-to-date.

**Comment by Greyisheep** [comment]
Thanks for the feedback.

**Addressing your remarks:**

> I think you are accidentally adding empty spaces, linting will fail
I will fix that right away

> can you explain what the addition in the release workflow `wheels.yml` will do?
The intention of the workflow there is to ensure that the "normal CI" is triggered if and only if a release is published (which was there previously), and the `Update Contributors` workflow is completed. My assumption of normal CI here is release and build wheels

Writing this now, I realized that what you referenced as "normal CI" in the issue, was a vaguely understood by me.

I then would like to ask, what particular CI workflow or action did you refer to as "normal CI" [here](https://github.com/sktime/sktime/issues/6181)

> I'd rather not couple the "update contributors" workflow to the release workflow, imo when the release button is being pushed, at that point we should assume that the contributors list is up-to-date.
`on` what would you like the update contributors to be triggered?
Not so sure what you mean here, but would removing the modification of wheels.yml and updating the name from `update_contributors.yml` to `00_update_contributors.yml`

**Comment by fkiraly** [comment]
> My assumption of normal CI here is release and build wheels

No, `wheels.yml` is only triggered for release.

The CI that runs on pull requests is `test.yml`.

**Comment by fkiraly** [comment]
Sure.

Currently, `update-contributors` and `test` lead to different workflow runs, e.g., here:
image

It seems that the `test` workflow triggers first, so the GitHub UI wil only register the latest workflow run for checking whether tests have passed on the PR, to validate whether required jobs have passed.

That is, while things like `test-nosoftdeps` run on the `test` action, the outcome becomes irrelevant for the purpose of informing the Github UI, because it's looking for the job in the `update-contributors` workflow.

Consequently, the jobs dashboard on the pull request looks like this:
image
even if all the jobs are run (but in the second-to-last workflow call form the PR), and no merge button appears.

If the `update-contributors` workflow were to run first, and then `test`, this would not be a problem - but it is unclear to me how to ensure that.

**Comment by sammychinedu2ky** [comment]
@fkiraly please where does it look for the job in the update-contributors workflow?

**Comment by fkiraly** [comment]
> @fkiraly please where does it look for the job in the update-contributors workflow?

 @sammychinedu2ky, could you kindly be more specific? what looks for which job?

**Comment by sammychinedu2ky** [comment]
> That is, while things like `test-nosoftdeps` run on the `test` action, the outcome becomes irrelevant for the purpose of informing the Github UI, because it's looking for the job in the `update-contributors` workflow.

You said this . so I am trying to understand what you meant by `looking for the job in the update-contributors` workflow

**Comment by fkiraly** [comment]
Ah, I see. The "looking for" is not part of the workflow, but part of the GitHub UI's behaviour, i.e., which yellow dots are displayed to turn into green tickmarks etc, below.

**Comment by sammychinedu2ky** [comment]
@fkiraly, why is it looking for the update-contributions workflow when it wasn't set to listen for a general push trigger? Also, I noticed that the flow requires approval from a maintainer to be executed. Maybe it will change to green after that.
image

**Comment by fkiraly** [comment]
this is a different issue - CI on this PR does not rune because of approval.

The issue with `update-contributors` only occurs when it is triggered, namely once the `all-contributorsrc` is changed (which currently does not happen in this PR).

**Comment by fkiraly** [comment]
The problem is, *if it runs*, then this happens:
https://github.com/sktime/sktime/pull/6189#issuecomment-2015095003

Let me know if you would like me to explain any part of the "what happens" in more detail.

**Comment by sammychinedu2ky** [comment]
> Consequently, the jobs dashboard on the pull request looks like this: image even if all the jobs are run (but in the second-to-last workflow call form the PR), and no merge button appears.
> 
> If the `update-contributors` workflow were to run first, and then `test`, this would not be a problem - but it is unclear to me how to ensure that.

@fkiraly  please can you explain this part

**Comment by fkiraly** [comment]
A *workflow run* consists of multiple *job runs*, you can inspect these when clicking "details". The *job runs* are those displayed in the GitHub dashboard. *workflow runs* and *job runs* are instances of the general *workflow* and *job* (specification) in the `yml`.

Does this help?

**Comment by sammychinedu2ky** [comment]
> A _workflow run_ consists of multiple _job runs_, you can inspect these when clicking "details". The _job runs_ are those displayed in the GitHub dashboard. _workflow runs_ and _job runs_ are instances of the general _workflow_ and _job_ (specification) in the `yml`.
> 
> Does this help?

Ooh I think I get it now. The problem is not related to this particular action for this pr. But the two workflow that will be triggered anytime a change is made to the contribution Json file right

**Comment by sammychinedu2ky** [comment]
Will adding an exclusion path to the contribution file fix this? @fkiraly

**Comment by fkiraly** [comment]
why would that help, what is your reasoning?

**Comment by sammychinedu2ky** [comment]
It will prevent a push to the all-contributorsrc from triggering the test workflow. An listener for a pull request can also be added

**Comment by fkiraly** [comment]
that wouldn't solve the issue, because we're looking at typical PR where both that file, and others are changed.

Quick question, @sammychinedu2ky, are you identical with @Greyisheep, or a different person?

**Comment by sammychinedu2ky** [comment]
😂 no I'm not. We are different @fkiraly

**Comment by fkiraly** [comment]
hm, ok, may I then suggest discussing further alternative approaches in the root issue #6181? Discussion in PR should be on the content of the PR, the code, primarily.

**Comment by Greyisheep** [comment]
> Sure.
> 
> Currently, `update-contributors` and `test` lead to different workflow runs, e.g., here: image
> 
> It seems that the `test` workflow triggers first, so the GitHub UI wil only register the latest workflow run for checking whether tests have passed on the PR, to validate whether required jobs have passed.
> 
> That is, while things like `test-nosoftdeps` run on the `test` action, the outcome becomes irrelevant for the purpose of informing the Github UI, because it's looking for the job in the `update-contributors` workflow.
> 
> Consequently, the jobs dashboard on the pull request looks like this: image even if all the jobs are run (but in the second-to-last workflow call form the PR), and no merge button appears.
> 
> If the `update-contributors` workflow were to run first, and then `test`, this would not be a problem - but it is unclear to me how to ensure that.

@fkiraly I think I will partition this into separate PRs: One for making sure update contributors doesn't run on main, and the other problem which is this back and forth with @sammychinedu2ky  can be resolved separately, so we can all brainstorm together.

**Comment by Greyisheep** [comment]
> that wouldn't solve the issue, because we're looking at typical PR where both that file, and others are changed.
> 
> Quick question, @sammychinedu2ky, are you identical with @Greyisheep, or a different person?

Different people

**Review by yarnabrina — APPROVED** [review]
Looks right.

It seems @duydl suggested this already, should we given them credit as well?

https://github.com/sktime/sktime/issues/6181#issuecomment-2012275116

**Comment by fkiraly** [comment]
> It seems @duydl suggested this already, should we given them credit as well?

Yes, that was one day earlier, agree. Will add him as a second contributor in the change log.

**Pull request merged**
